### PR TITLE
[MLMC] New application structure

### PR DESCRIPTION
**Pull request by riccardotosi**
This PR's aim is to redefine the structure of the application. The structure is the following:

* python_scripts: contains the utilities to perform the Monte Carlo and Multilevel Monte Carlo algorithms. The test_*.py files are needed to run the application tests without COMPSs,
* tests: contains the tests of the application. In a successive PR tests for specific function of the application (such as the computation of central moments) will be considered. It contains folders with the corresponding application-test, in this case we have the "poisson_square_2d" folder.
    * poisson_square_2d: this folder contains the analysis stage of the problem, which is derived from the specific application we are considering, and a main, that is the file to run in order to execute the simulation,
        * problem_settings: contains model part and project parameters of each application, the settings of the monte carlo simulation and of the refinement in case of multilevel monte carlo simulation.

The analysis stage `SimulationScenario` is what the user should modify in order to solve a different problem, while all the utilities inside python_scripts and the main file should remain the same.
The application works also using PyCOMPSs.

In the Examples repository we will add in future specific problems (such as compressible potential flow) that are not contained inside the tests folder.

The flag `store_lower_levels_samples` is added, and gives the possibility to the user to decide if storing or not the samples computed through the adaptive refinement.

### CODE REVIEW

File: kratos/python_scripts/KratosUnittest.py
Comment (philbucher) [inline kratos/python_scripts/KratosUnittest.py]: changes to the core (esp such a class) affect everyone, please make a separate PR explaining why the changes are needed

btw I am not sure if this can go in, we currently try to remove adding things to the path bcs it messes things up
=> to be discussed in a separate PR

**Review by philbucher — CHANGES_REQUESTED** [review]

File: kratos/python_scripts/KratosUnittest.py
Comment (riccardotosi) [reply kratos/python_scripts/KratosUnittest.py]: I discussed with @roigcarlo and he proposed to put this change in this specific PR since it is needed in this specific PR.

Comment (roigcarlo) [reply kratos/python_scripts/KratosUnittest.py]: @philbucher As @riccardotosi says this is somehow hard to understand without an specific example which needs to use this functionality. Hance why is in here. but if you preffer we candiscuss it in a separate PR.

But as the topic has already appeared, a little explanation is in order:
The problem here is caused because the specific example we need to run assumes that, apart from the regular MainKratos.py, the directory from which the problem will run has an aditional *.py file. Unfortunatelly right now there is no other way to proced with this. 

Normally this would not be problem, but since the Unittest library runs the tests from an arbitrary palce we need to tell python where to look for this file. Since we already have a function which manipulates the dir it seemed logic to put it here.

Comment (philbucher) [reply kratos/python_scripts/KratosUnittest.py]: no worries @roigcarlo :)

Why don't you manually add the path of the test-file to the python-path in the respective test? This should work if the "other" py-file is in the same directory

I am just afraid if we add it by default then we could hide some import-problems :/

=> if my first proposal doesn't work then could you make the adding to the path an optional thing? I.e. a constructor argument defaulting to `False`?

Comment (roigcarlo) [reply kratos/python_scripts/KratosUnittest.py]: I like more the second option, maybe we can make a separate context manager just for paths as we have for dirs, I didn't think on that one honestly

Comment (philbucher) [reply kratos/python_scripts/KratosUnittest.py]: @roigcarlo up to you :) 
I think here a default arg could make more sense, since the rest of the contextmanager is the same (as far as I understood) 
=> less chances of the implementations to diverge in the future

@riccardotosi I will remove my block as soon as someone reviewed the rest, since it is currently not possible to dismiss reviews and I cannot approve this

**Comment by riccardotosi** [comment]
Thanks @roigcarlo!

**Comment by philbucher** [comment]
just to say that for me the changes in the Core are ok
I will release my block as soon as someone else approves this PR, since I cannot approve it

**Comment by riccardotosi** [comment]
@RiccardoRossi ping

File: applications/MultilevelMonteCarloApplication/python_scripts/mc_utilities.py
Comment (philbucher) [inline applications/MultilevelMonteCarloApplication/python_scripts/mc_utilities.py]: side note: this way of importing python-scripts is deprecated and will stop working at some point (not soon though) therefore we encourage devs to slowly adapt the new way of importing:
```suggestion
from KratosMultiphysics.MultilevelMonteCarloApplication.statistical_variable_utilities import StatisticalVariable
```

**Review by philbucher — DISMISSED** [review]

File: applications/MultilevelMonteCarloApplication/python_scripts/mc_utilities.py
Comment (riccardotosi) [reply applications/MultilevelMonteCarloApplication/python_scripts/mc_utilities.py]: Thanks

Comment (philbucher) [reply applications/MultilevelMonteCarloApplication/python_scripts/mc_utilities.py]: You can check the CompressiblePotentialApp, they already use the [new import system](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/CompressiblePotentialFlowApplication/CompressiblePotentialFlowApplication.py#L11)

Comment (riccardotosi) [reply applications/MultilevelMonteCarloApplication/python_scripts/mc_utilities.py]: Ty, done!

File: applications/MultilevelMonteCarloApplication/MultilevelMonteCarloApplication.py
Comment (philbucher) [inline applications/MultilevelMonteCarloApplication/MultilevelMonteCarloApplication.py]:
```suggestion
import KratosMultiphysics as KM
```

File: applications/MultilevelMonteCarloApplication/MultilevelMonteCarloApplication.py
Comment (philbucher) [inline applications/MultilevelMonteCarloApplication/MultilevelMonteCarloApplication.py:11]:
```suggestion
KM._ImportApplicationAsModule(application, application_name, application_folder, __path__)
```

**Review by philbucher — COMMENTED** [review]
Nice!

If imports of some python-scripts stop working, just add `KratosMultiphysics.MultilevelMonteCarloApplication.` before the name of the python-script, then it should work again

**Comment by pablobecker** [comment]
In principle, the @KratosMultiphysics/implementation-committee has no problems with yourchanges to the core. The rest looks opk too. However, we must ask you to separate this single change to the core and put it in its own PR so we can more cleanly evaluate it.

**Comment by philbucher** [comment]
@riccardotosi I recommend to take a look at codacy
Some of the suggestions are relevant, esp the ones abt unused imports

**Comment by roigcarlo** [comment]
Just passing by to tell that probably is not a good idea to blindly accept codacy suggestions about unused imports in this particular scenario. I can't tell you with a 100% confidence all of the reports are wrong but I suspect some of them might be.

I explain myself. The python code of this PR is not meant to be executed by a "regular" python runtime, but rather by a swarm of them. The problem here appears while constructing the communication of data between instances which is done by pickle and unpickle the python classes. 
The problem is that pickle does not provide a mechanism to store the class "prototype" (let me use js jargon here as I don't know the python equivalent).

With a little example, imagine these scripts:
```python
# Myexecutionbody.py
import pycompps

@pycompss_task
def distributed_task(pickled_object):
     myclass_instance = unpickle(pickled_object)
     print(myclass_instance.id)
```

```python
 # Main.py (pseudocode)
import myclass
import myexecutionbody

myclass_pickled = pickle(myclass("NewInstance"))

myexecutionbody.distributed_task(myclass_pickled)
```

Without entering in the details of the implementation of the pickled class, you can see that if I try to execute the `distributed_task` without having been called by the `Main.py` it is impossible to restore the object as I don't know what it is. That situation should be impossible if running python from your commandline but its its possible with pyCompss.

It happens something similar with the kratos file serialization. If I serialize something into a file which contains classes from the StructuralMechanics, and then I try to load it in another script which only loads the FluidDynamics its going to error.

File: applications/MultilevelMonteCarloApplication/MultilevelMonteCarloApplication.py
Comment (riccardotosi) [reply applications/MultilevelMonteCarloApplication/MultilevelMonteCarloApplication.py:11]: Done

**Comment by AlejandroCornejo** [comment]
Since the changes in the core have been removed in this PR, the @KratosMultiphysics/implementation-committee approves the changes (@adityaghantasala has already expressed his agreement based on the this change).  Waiting for the final ok from the @KratosMultiphysics/technical-committee

**Comment by riccardotosi** [comment]
@RiccardoRossi ping

**Review by roigcarlo — APPROVED** [review]
All good from the @KratosMultiphysics/technical-committee. Sorry for taking that long to review

**Comment by RiccardoRossi** [comment]
+1!!!

**Pull request merged**
